## Algoritmos de búsqueda en Torre de Hanoi

## Inicialización del problema

In [ ]:
# Número de discos a utilizar
disks = 5

In [ ]:
# Lista con el número de discos a utilizar
list = [x for x in range(disks, 0, -1)]

In [ ]:
from hanoi_states import ProblemHanoi, StatesHanoi, ActionHanoi
from tree_hanoi import NodeHanoi
from aima import PriorityQueue
import tracemalloc


In [ ]:
initial_state = StatesHanoi(list, [], [], max_disks=disks)
goal_state = StatesHanoi([], [], list, max_disks=disks)
problem = ProblemHanoi(initial=initial_state, goal=goal_state)

## Búsqueda primero en anchura

In [ ]:
def bfs(p):
    """
    Algoritmo de búsqueda no informada.
    Se expande primero el nodo raíz, a continuación,
    se expanden todos los sucesores del nodo raíz,
    después sus sucesores, etc.

    Args:
        HanoiProblem

    Returns:
        Nodo resultado, número de nodos explorados.
    """
    frontier = [NodeHanoi(p.initial)]
    explored = set()

    while len(frontier) != 0:
        node = frontier.pop()
        explored.add(node.state)

        if p.goal_test(node.state):
            last_node = node
            break

        for next_node in node.expand(p):
            if next_node.state not in explored:
                frontier.insert(0, next_node)

    return last_node, len(explored)

In [ ]:
%%timeit
bfs(problem)

80.3 ms ± 1.78 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


### Resultados - Primero en anchura

In [ ]:
tracemalloc.start()
bfs_result_node, bfs_explored = bfs(problem)
_, memory_peak = tracemalloc.get_traced_memory()
memory_peak /= 1024*1024
tracemalloc.stop()

In [ ]:
print(f"Memoria máxima: {round(memory_peak, 2)} MB.", )
totalcost=round(bfs_result_node.state.accumulated_cost)
print(f'Pasos hasta encontrar la solución: {totalcost}.')
print(f'Cantidad de nodos explorados: {bfs_explored}.')

node = bfs_result_node
result = []
print('Detalle de movimientos:')
while node.parent is not None:
    result.append(node.state)
    node = node.parent
result.append(node.state)

i=0
for x in range(totalcost, -1, -1):
    print (f'{i}) {result[x]}')
    i+=1

Memoria máxima: 1.38 MB.
Pasos hasta encontrar la solución: 31.
Cantidad de nodos explorados: 233.
Detalle de movimientos:
0) HanoiState: 5 4 3 2 1 |  | 
1) HanoiState: 5 4 3 2 |  | 1
2) HanoiState: 5 4 3 | 2 | 1
3) HanoiState: 5 4 3 | 2 1 | 
4) HanoiState: 5 4 | 2 1 | 3
5) HanoiState: 5 4 1 | 2 | 3
6) HanoiState: 5 4 1 |  | 3 2
7) HanoiState: 5 4 |  | 3 2 1
8) HanoiState: 5 | 4 | 3 2 1
9) HanoiState: 5 | 4 1 | 3 2
10) HanoiState: 5 2 | 4 1 | 3
11) HanoiState: 5 2 1 | 4 | 3
12) HanoiState: 5 2 1 | 4 3 | 
13) HanoiState: 5 2 | 4 3 | 1
14) HanoiState: 5 | 4 3 2 | 1
15) HanoiState: 5 | 4 3 2 1 | 
16) HanoiState:  | 4 3 2 1 | 5
17) HanoiState: 1 | 4 3 2 | 5
18) HanoiState: 1 | 4 3 | 5 2
19) HanoiState:  | 4 3 | 5 2 1
20) HanoiState: 3 | 4 | 5 2 1
21) HanoiState: 3 | 4 1 | 5 2
22) HanoiState: 3 2 | 4 1 | 5
23) HanoiState: 3 2 1 | 4 | 5
24) HanoiState: 3 2 1 |  | 5 4
25) HanoiState: 3 2 |  | 5 4 1
26) HanoiState: 3 | 2 | 5 4 1
27) HanoiState: 3 | 2 1 | 5 4
28) HanoiState:  | 2 1 | 5 4 3
29) 

## Búsqueda primero en profundidad

In [ ]:
def dfs(p):
    """
    Algoritmo de búsqueda no informada.
    Siempre expande el nodo más profundo en la frontera actual del árbol de búsqueda.
    Cuando esos nodos se expanden, son quitados de la frontera,
    así entonces la búsqueda “retrocede” al siguiente nodo más superficial.

    Args:
        HanoiProblem

    Returns:
        Nodo resultado, número de nodos explorados.
    """

    frontier = [NodeHanoi(p.initial)]
    explored = set()

    while len(frontier) != 0:
        node = frontier.pop()
        explored.add(node.state)

        if p.goal_test(node.state):
            last_node = node
            break

        for next_node in node.expand(p):
            if next_node.state not in explored:
                frontier.append(next_node)

    return last_node, len(explored)

In [ ]:
%%timeit
dfs(problem)

6.9 ms ± 150 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


### Resultados - Primero en profundidad

In [ ]:
tracemalloc.start()
dfs_result_node, dfs_explored = dfs(problem)
_, memory_peak = tracemalloc.get_traced_memory()
memory_peak /= 1024*1024
tracemalloc.stop()

In [ ]:
print(f"Memoria máxima: {round(memory_peak, 2)} MB.", )
totalcost=round(dfs_result_node.state.accumulated_cost)
print(f'Pasos hasta encontrar la solución: {totalcost}.')
print(f'Cantidad de nodos explorados: {dfs_explored}.')

node = dfs_result_node
result = []
print('Detalle de movimientos:')
while node.parent is not None:
    result.append(node.state)
    node = node.parent
result.append(node.state)

i=0
for x in range(totalcost, -1, -1):
    print (f'{i}) {result[x]}')
    i+=1

Memoria máxima: 0.21 MB.
Pasos hasta encontrar la solución: 121.
Cantidad de nodos explorados: 122.
Detalle de movimientos:
0) HanoiState: 5 4 3 2 1 |  | 
1) HanoiState: 5 4 3 2 |  | 1
2) HanoiState: 5 4 3 2 | 1 | 
3) HanoiState: 5 4 3 | 1 | 2
4) HanoiState: 5 4 3 |  | 2 1
5) HanoiState: 5 4 3 1 |  | 2
6) HanoiState: 5 4 3 1 | 2 | 
7) HanoiState: 5 4 3 | 2 | 1
8) HanoiState: 5 4 3 | 2 1 | 
9) HanoiState: 5 4 | 2 1 | 3
10) HanoiState: 5 4 | 2 | 3 1
11) HanoiState: 5 4 1 | 2 | 3
12) HanoiState: 5 4 1 |  | 3 2
13) HanoiState: 5 4 |  | 3 2 1
14) HanoiState: 5 4 | 1 | 3 2
15) HanoiState: 5 4 2 | 1 | 3
16) HanoiState: 5 4 2 |  | 3 1
17) HanoiState: 5 4 2 1 |  | 3
18) HanoiState: 5 4 2 1 | 3 | 
19) HanoiState: 5 4 2 | 3 | 1
20) HanoiState: 5 4 2 | 3 1 | 
21) HanoiState: 5 4 | 3 1 | 2
22) HanoiState: 5 4 | 3 | 2 1
23) HanoiState: 5 4 1 | 3 | 2
24) HanoiState: 5 4 1 | 3 2 | 
25) HanoiState: 5 4 | 3 2 | 1
26) HanoiState: 5 4 | 3 2 1 | 
27) HanoiState: 5 | 3 2 1 | 4
28) HanoiState: 5 | 3 2 | 4 1


## Búsqueda de costo uniforme

In [ ]:
def ucs(p):
    """
    Algoritmo de búsqueda no informada.
    Expande el nodo con el camino de costo más pequeño.

    Args:
        HanoiProblem

    Returns:
        Nodo resultado, número de nodos explorados.
    """
    def path_cost(item):
        return item.path_cost

    frontier = PriorityQueue(order='min', f=path_cost)

    frontier.append(NodeHanoi(p.initial))
    explored = set()

    while len(frontier) != 0:
        node = frontier.pop()
        explored.add(node.state)

        if p.goal_test(node.state):
            last_node = node
            break

        for next_node in node.expand(p):
            if next_node.state not in explored:
                frontier.append(next_node)

    return last_node, len(explored)

In [ ]:
%%timeit
ucs(problem)

24.2 ms ± 792 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)


### Resultados - Costo uniforme

In [ ]:
tracemalloc.start()
ucs_result_node, ucs_explored = ucs(problem)
_, memory_peak = tracemalloc.get_traced_memory()
memory_peak /= 1024*1024
tracemalloc.stop()

In [ ]:
print(f"Memoria máxima: {round(memory_peak, 2)} MB.", )
totalcost=round(ucs_result_node.state.accumulated_cost)
print(f'Pasos hasta encontrar la solución: {totalcost}.')
print(f'Cantidad de nodos explorados: {ucs_explored}.')

node = ucs_result_node
result = []
print('Detalle de movimientos:')
while node.parent is not None:
    result.append(node.state)
    node = node.parent
result.append(node.state)

i=0
for x in range(totalcost, -1, -1):
    print (f'{i}) {result[x]}')
    i+=1

Memoria máxima: 0.28 MB.
Pasos hasta encontrar la solución: 31.
Cantidad de nodos explorados: 216.
Detalle de movimientos:
0) HanoiState: 5 4 3 2 1 |  | 
1) HanoiState: 5 4 3 2 |  | 1
2) HanoiState: 5 4 3 | 2 | 1
3) HanoiState: 5 4 3 | 2 1 | 
4) HanoiState: 5 4 | 2 1 | 3
5) HanoiState: 5 4 1 | 2 | 3
6) HanoiState: 5 4 1 |  | 3 2
7) HanoiState: 5 4 |  | 3 2 1
8) HanoiState: 5 | 4 | 3 2 1
9) HanoiState: 5 | 4 1 | 3 2
10) HanoiState: 5 2 | 4 1 | 3
11) HanoiState: 5 2 1 | 4 | 3
12) HanoiState: 5 2 1 | 4 3 | 
13) HanoiState: 5 2 | 4 3 | 1
14) HanoiState: 5 | 4 3 2 | 1
15) HanoiState: 5 | 4 3 2 1 | 
16) HanoiState:  | 4 3 2 1 | 5
17) HanoiState: 1 | 4 3 2 | 5
18) HanoiState: 1 | 4 3 | 5 2
19) HanoiState:  | 4 3 | 5 2 1
20) HanoiState: 3 | 4 | 5 2 1
21) HanoiState: 3 | 4 1 | 5 2
22) HanoiState: 3 2 | 4 1 | 5
23) HanoiState: 3 2 1 | 4 | 5
24) HanoiState: 3 2 1 |  | 5 4
25) HanoiState: 3 2 |  | 5 4 1
26) HanoiState: 3 | 2 | 5 4 1
27) HanoiState: 3 | 2 1 | 5 4
28) HanoiState:  | 2 1 | 5 4 3
29) 

## Búsqueda de profundidad limitada

In [ ]:
# Parámetro - Límite de niveles de profundidad
hyperparam=200

In [ ]:
def dfs_l(p, limit):
    """
    Algoritmo de búsqueda no informada.
    Limita la búsqueda en profundidad hasta un el nivel límite.

    Args:
        Límite de profundidad

    Returns:
        Nodo resultado, número de nodos explorados.
    """

    explored = set()

    def recursive_dfs(node, p, limit):
        nonlocal explored

        if p.goal_test(node.state):
            return node
        elif limit==0:
            return 'cutoff'
        else:
            cutoff = False
            for child in node.expand(p):
                s = child.state
                if s not in explored:
                    explored.add(s)
                    result = recursive_dfs(child, p, limit - 1)

                    if result == 'cutoff':
                        cutoff = True
                    elif result is not None:
                        return result
        if cutoff:
            return 'cutoff'
        return None
    return recursive_dfs(NodeHanoi(p.initial), p, limit), len(explored)

In [ ]:
%%timeit
dfs_l(problem, hyperparam)

10.8 ms ± 135 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


### Resultados - Profundidad Limitada

In [ ]:
tracemalloc.start()
dfs_l_result_node, dfs_l_explored = dfs_l(problem, hyperparam)
_, memory_peak = tracemalloc.get_traced_memory()
memory_peak /= 1024*1024
tracemalloc.stop()

In [ ]:
print(f"Memoria máxima: {round(memory_peak, 2)} MB.", )
print(f'Cantidad de nodos explorados: {dfs_l_explored}.')

if(dfs_l_result_node == 'cutoff'):
    print('NO ENCONTRÓ SOLUCIÓN')
else:
    totalcost=round(dfs_l_result_node.state.accumulated_cost)
    print(f'Pasos hasta encontrar la solución: {totalcost}.')

    node = dfs_l_result_node
    result = []
    print('Detalle de movimientos:')
    while node.parent is not None:
        result.append(node.state)
        node = node.parent
    result.append(node.state)

    i=0
    for x in range(totalcost, -1, -1):
        print (f'{i}) {result[x]}')
        i+=1

Memoria máxima: 0.43 MB.
Cantidad de nodos explorados: 162.
Pasos hasta encontrar la solución: 122.
Detalle de movimientos:
0) HanoiState: 5 4 3 2 1 |  | 
1) HanoiState: 5 4 3 2 | 1 | 
2) HanoiState: 5 4 3 | 1 | 2
3) HanoiState: 5 4 3 1 |  | 2
4) HanoiState: 5 4 3 |  | 2 1
5) HanoiState: 5 4 | 3 | 2 1
6) HanoiState: 5 4 1 | 3 | 2
7) HanoiState: 5 4 | 3 1 | 2
8) HanoiState: 5 4 2 | 3 1 | 
9) HanoiState: 5 4 2 1 | 3 | 
10) HanoiState: 5 4 2 | 3 | 1
11) HanoiState: 5 4 | 3 2 | 1
12) HanoiState: 5 4 1 | 3 2 | 
13) HanoiState: 5 4 | 3 2 1 | 
14) HanoiState: 5 | 3 2 1 | 4
15) HanoiState: 5 1 | 3 2 | 4
16) HanoiState: 5 | 3 2 | 4 1
17) HanoiState: 5 2 | 3 | 4 1
18) HanoiState: 5 2 1 | 3 | 4
19) HanoiState: 5 2 | 3 1 | 4
20) HanoiState: 5 | 3 1 | 4 2
21) HanoiState: 5 1 | 3 | 4 2
22) HanoiState: 5 | 3 | 4 2 1
23) HanoiState: 5 3 |  | 4 2 1
24) HanoiState: 5 3 1 |  | 4 2
25) HanoiState: 5 3 | 1 | 4 2
26) HanoiState: 5 3 2 | 1 | 4
27) HanoiState: 5 3 2 1 |  | 4
28) HanoiState: 5 3 2 |  | 4 1
29)

## Búsqueda de profundidad limitada iterativa

In [ ]:
def dfs_l_i(p):
    """
    Algoritmo de búsqueda no informada.
    Va aumentando la profundidad iterativamente hasta alcanzar la solución.

    Args:
        HanoiProblem

    Returns:
        Nodo resultado, número de nodos explorados.
    """

    depth = 0
    while True:
        r, e = dfs_l(p, depth)
        depth+=1
        if r != 'cutoff':
            break

    return r, e



In [ ]:
%%timeit
dfs_l_i(problem)

387 ms ± 11.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


### Resultados - Profundidad limitada iterativa

In [ ]:
tracemalloc.start()
dfs_l_i_result_node, dfs_l_i_result_explored = dfs_l_i(problem)
_, memory_peak = tracemalloc.get_traced_memory()
memory_peak /= 1024*1024
tracemalloc.stop()

In [ ]:
print(f"Memoria máxima: {round(memory_peak, 2)} MB.", )
totalcost=round(dfs_l_i_result_node.state.accumulated_cost)
print(f'Pasos hasta encontrar la solución: {totalcost}.')
print(f'Cantidad de nodos explorados: {dfs_l_i_result_explored}.')

node = dfs_l_i_result_node
result = []
print('Detalle de movimientos:')
while node.parent is not None:
    result.append(node.state)
    node = node.parent
result.append(node.state)

i=0
for x in range(totalcost, -1, -1):
    print (f'{i}) {result[x]}')
    i+=1

Memoria máxima: 1.19 MB.
Pasos hasta encontrar la solución: 65.
Cantidad de nodos explorados: 164.
Detalle de movimientos:
0) HanoiState: 5 4 3 2 1 |  | 
1) HanoiState: 5 4 3 2 | 1 | 
2) HanoiState: 5 4 3 | 1 | 2
3) HanoiState: 5 4 3 1 |  | 2
4) HanoiState: 5 4 3 |  | 2 1
5) HanoiState: 5 4 | 3 | 2 1
6) HanoiState: 5 4 1 | 3 | 2
7) HanoiState: 5 4 | 3 1 | 2
8) HanoiState: 5 4 2 | 3 1 | 
9) HanoiState: 5 4 2 1 | 3 | 
10) HanoiState: 5 4 2 | 3 | 1
11) HanoiState: 5 4 | 3 2 | 1
12) HanoiState: 5 4 1 | 3 2 | 
13) HanoiState: 5 4 | 3 2 1 | 
14) HanoiState: 5 | 3 2 1 | 4
15) HanoiState: 5 1 | 3 2 | 4
16) HanoiState: 5 | 3 2 | 4 1
17) HanoiState: 5 2 | 3 | 4 1
18) HanoiState: 5 2 1 | 3 | 4
19) HanoiState: 5 2 | 3 1 | 4
20) HanoiState: 5 | 3 1 | 4 2
21) HanoiState: 5 1 | 3 | 4 2
22) HanoiState: 5 | 3 | 4 2 1
23) HanoiState: 5 3 |  | 4 2 1
24) HanoiState: 5 3 1 |  | 4 2
25) HanoiState: 5 3 | 1 | 4 2
26) HanoiState: 5 3 2 | 1 | 4
27) HanoiState: 5 3 2 1 |  | 4
28) HanoiState: 5 3 2 1 | 4 | 
29) 

## Búsqueda voraz

In [ ]:
def greedy(p):
    """
    Algoritmo de búsqueda informada.
    Trata de expandir el nodo con el valor más bajo de la función heurística
    (el nodo que parece más cerca del objetivo).

    Args:
        HanoiProblem

    Returns:
        Nodo resultado, número de nodos explorados.
    """

    node = NodeHanoi(p.initial)
    explored = set()

    def hfunction(item):
        nonlocal p
        hcost = 0

        for x,y in zip(item.state.get_state()[-1], p.goal.get_state()[-1]):
            if x==y:
                hcost = hcost -1
        return hcost

    frontier = PriorityQueue(order='min', f=hfunction)
    frontier.append(node)

    if p.goal_test(node.state):
        return node, len(explored)

    while len(frontier) != 0:
        node = frontier.pop()

        for next_node in node.expand(p):

            if p.goal_test(next_node.state):
                return next_node, len(explored)

            if next_node.state not in explored:
                explored.add(next_node.state)
                frontier.append(next_node)

            elif next_node in frontier:
                if hfunction(next_node) < frontier[next_node]:
                    del frontier[next_node]
                    frontier.append(next_node)
    return "ERROR"

In [ ]:
%%timeit
greedy(problem)

7.61 ms ± 237 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


### Resultados - Greedy

In [ ]:
tracemalloc.start()
greedy_result_node, greedy_explored = greedy(problem)
_, memory_peak = tracemalloc.get_traced_memory()
memory_peak /= 1024*1024
tracemalloc.stop()

In [ ]:
print(f"Memoria máxima: {round(memory_peak, 2)} MB.", )
totalcost=round(greedy_result_node.state.accumulated_cost)
print(f'Pasos hasta encontrar la solución: {totalcost}.')
print(f'Cantidad de nodos explorados: {greedy_explored}.')

node = greedy_result_node
result = []
print('Detalle de movimientos:')
while node.parent is not None:
    result.append(node.state)
    node = node.parent
result.append(node.state)

i=0
for x in range(totalcost, -1, -1):
    print (f'{i}) {result[x]}')
    i+=1

Memoria máxima: 0.13 MB.
Pasos hasta encontrar la solución: 31.
Cantidad de nodos explorados: 123.
Detalle de movimientos:
0) HanoiState: 5 4 3 2 1 |  | 
1) HanoiState: 5 4 3 2 |  | 1
2) HanoiState: 5 4 3 | 2 | 1
3) HanoiState: 5 4 3 | 2 1 | 
4) HanoiState: 5 4 | 2 1 | 3
5) HanoiState: 5 4 1 | 2 | 3
6) HanoiState: 5 4 1 |  | 3 2
7) HanoiState: 5 4 |  | 3 2 1
8) HanoiState: 5 | 4 | 3 2 1
9) HanoiState: 5 | 4 1 | 3 2
10) HanoiState: 5 2 | 4 1 | 3
11) HanoiState: 5 2 1 | 4 | 3
12) HanoiState: 5 2 1 | 4 3 | 
13) HanoiState: 5 2 | 4 3 | 1
14) HanoiState: 5 | 4 3 2 | 1
15) HanoiState: 5 | 4 3 2 1 | 
16) HanoiState:  | 4 3 2 1 | 5
17) HanoiState: 1 | 4 3 2 | 5
18) HanoiState: 1 | 4 3 | 5 2
19) HanoiState:  | 4 3 | 5 2 1
20) HanoiState: 3 | 4 | 5 2 1
21) HanoiState: 3 | 4 1 | 5 2
22) HanoiState: 3 2 | 4 1 | 5
23) HanoiState: 3 2 1 | 4 | 5
24) HanoiState: 3 2 1 |  | 5 4
25) HanoiState: 3 2 |  | 5 4 1
26) HanoiState: 3 | 2 | 5 4 1
27) HanoiState: 3 | 2 1 | 5 4
28) HanoiState:  | 2 1 | 5 4 3
29) 

## Búsqueda A*

In [ ]:
def a_star(p):
    """
    Algoritmo de búsqueda informada.
    Usa la función heurística más el costo del camino
    tomado para llegar al nodo.

    Args:
        HanoiProblem

    Returns:
        Nodo resultado, número de nodos explorados.
    """

    node = NodeHanoi(p.initial)
    explored = set()

    def hfunction(item):
        nonlocal p
        hcost = 0

        for x,y in zip(item.state.get_state()[-1], p.goal.get_state()[-1]):
            if x==y:
                hcost = hcost -1
        return hcost + item.path_cost

    frontier = PriorityQueue(order='min', f=hfunction)
    frontier.append(node)

    if p.goal_test(node.state):
        return node, len(explored)

    while len(frontier) != 0:
        node = frontier.pop()

        for next_node in node.expand(p):

            if p.goal_test(next_node.state):
                return next_node, len(explored)

            if next_node.state not in explored:
                explored.add(next_node.state)
                frontier.append(next_node)

            elif next_node in frontier:
                if hfunction(next_node) < frontier[next_node]:
                    del frontier[next_node]
                    frontier.append(next_node)
    return "ERROR"

In [ ]:
%%timeit
a_star(problem)

12.4 ms ± 389 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


### Resultados - A*

In [ ]:
tracemalloc.start()
a_star_result_node, a_star_explored = a_star(problem)
_, memory_peak = tracemalloc.get_traced_memory()
memory_peak /= 1024*1024
tracemalloc.stop()

In [ ]:
print(f"Memoria máxima: {round(memory_peak, 2)} MB.", )
totalcost=round(a_star_result_node.state.accumulated_cost)
print(f'Pasos hasta encontrar la solución: {totalcost}.')
print(f'Cantidad de nodos explorados: {a_star_explored}.')

node = a_star_result_node
result = []
print('Detalle de movimientos:')
while node.parent is not None:
    result.append(node.state)
    node = node.parent
result.append(node.state)

i=0
for x in range(totalcost, -1, -1):
    print (f'{i}) {result[x]}')
    i+=1

Memoria máxima: 0.17 MB.
Pasos hasta encontrar la solución: 31.
Cantidad de nodos explorados: 182.
Detalle de movimientos:
0) HanoiState: 5 4 3 2 1 |  | 
1) HanoiState: 5 4 3 2 |  | 1
2) HanoiState: 5 4 3 | 2 | 1
3) HanoiState: 5 4 3 | 2 1 | 
4) HanoiState: 5 4 | 2 1 | 3
5) HanoiState: 5 4 1 | 2 | 3
6) HanoiState: 5 4 1 |  | 3 2
7) HanoiState: 5 4 |  | 3 2 1
8) HanoiState: 5 | 4 | 3 2 1
9) HanoiState: 5 | 4 1 | 3 2
10) HanoiState: 5 2 | 4 1 | 3
11) HanoiState: 5 2 1 | 4 | 3
12) HanoiState: 5 2 1 | 4 3 | 
13) HanoiState: 5 2 | 4 3 | 1
14) HanoiState: 5 | 4 3 2 | 1
15) HanoiState: 5 | 4 3 2 1 | 
16) HanoiState:  | 4 3 2 1 | 5
17) HanoiState: 1 | 4 3 2 | 5
18) HanoiState: 1 | 4 3 | 5 2
19) HanoiState:  | 4 3 | 5 2 1
20) HanoiState: 3 | 4 | 5 2 1
21) HanoiState: 3 | 4 1 | 5 2
22) HanoiState: 3 2 | 4 1 | 5
23) HanoiState: 3 2 1 | 4 | 5
24) HanoiState: 3 2 1 |  | 5 4
25) HanoiState: 3 2 |  | 5 4 1
26) HanoiState: 3 | 2 | 5 4 1
27) HanoiState: 3 | 2 1 | 5 4
28) HanoiState:  | 2 1 | 5 4 3
29) 

## Conclusiones



Algoritmos que encontraron la solución óptima:
Para 5 discos, 2⁵-1=31 pasos.
- Primero en anchura
- Costo uniforme
- Greedy
- A*

De los algoritmos que no encontraron solución óptima, cuáles tuvieron mejor performance (en términos de pasos para llegar a la solución):
1) Profundidad limitada iterativa  
2) Primero en profundidad  
  
Algoritmos ordenados por tiempo de respuesta obtenido:  
1) Primero en profundidad (6.95 ms)  
2) Greedy (7.36 ms)  
2) A* (11.9 ms)  
4) Costo uniforme (24.3 ms)  
5) Primero en anchura (80.3 ms)  
6) Profundidad limitada iterativa (391 ms)  
  
Algoritmos ordenados por cantidad de memoria consumida:  
1) Greedy (0.13 MB)  
2) A* (0.17 MB)  
3) Primero en profundidad (0.21 MB)  
4) Costo uniforme (0.28 MB)  
5) Profundidad limitada iterativa (1.19 MB)  
6) Primero en anchura (1.38 MB)  
  
Algoritmos ordenados por cantidad de nodos explorados:  
1) Primero en profundidad (122)  
2) Greedy (123)  
3) Profundidad limitada iterativa (164)  
4) A* (182)  
5) Costo uniforme (216)  
6) Primero en anchura (233)  





### Orden de Complejidad de los algoritmos



Siendo

b: factor de ramificación en el espacio de estados.

d: profundidad de la solución.

m: profundidad máxima de cualquier camino en el espacio de estados.


* La búsqueda primero en anchura selecciona para su expansión el nodo no expandido más superficial en el árbol de búsqueda.

Complejidad en tiempo: O(b^d)

Complejidad en espacio: O(b^d)


* La búsqueda primero en profundidad selecciona para la expansión el nodo no expandido más profundo del árbol de búsqueda.

    Complejidad en tiempo: O(b^m)

    Complejidad en espacio: O(bm)

* La búsqueda de profundidad iterativa llama a la búsqueda de profundidad limitada aumentando este límite hasta que se encuentre el objetivo.

    Complejidad en tiempo: O(b^d)

    Complejidad en espacio: O(bd)


* La búsqueda voraz trata de expandir el nodo más cercano al objetivo alegando que probablemente conduzca rápidamente a una solución.

    Complejidad en tiempo(*): O(b^m)

    Complejidad en espacio(*): O(b^m)

    (*) Corresponden al peor de los casos, se puede reducir la complejidad con una mejor calidad de la heurística.


* La búsqueda A-estrella evalúa los nodos conbinando la función heurística y el coste de ir al nodo objetivo.

    El rendimiento de estos algoritmos depende de la calidad de la función heurística.
    La complejidad en tiempo puede ir del orden lineal al exponencial.
    La complejidad en espacio puede llegar también a ser exponencial, para esto existen algunas variantes.
    Por ejemplo, el A*M (A* con memoria acotada) y A*MS (A*M simplificado) que utilizan una cantidad limitada de memoria.



## Bibliografía


Inteligencia Artificial - Un Enfoque Moderno (2da.Ed.) - Stuart Russell, Peter Norvig.

